docs/language.md

```diff
@@ -6604,37 +6852,128 @@ TODO
 
 -->
 
-## Built-in Functions
+## Run-time Types
+
+Types can be represented at run-time.
+To create a type value, use the constructor function `Type<T>()`, which accepts the static type as a type argument.
+
+This is similar to e.g. `T.self` in Swift, `T::class` in Kotlin, and `T.class` in Java.
 
-### `panic`
+For example, to represent the type `Int` at run-time:
 
 ```cadence
-fun panic(_ message: String): Never
+let intType: Type = Type<Int>()
 ```
 
-Terminates the program unconditionally
-and reports a message which explains why the unrecoverable error occurred.
+This works for both built-in and user-defined types. For example, to get the type value for a resource:
 
-#### Example
+```cadence,file=runtime-type-collectible.cdc
+resource Collectible {}
 
-```cadence
-let optionalAccount: Account? = // ...
-let account = optionalAccount ?? panic("missing account")
+let collectibleType = Type<@Collectible>()
+
+// `collectibleType` has type `Type`
+```
+
+Type values are comparable.
+
+```cadence,file=runtime-type-comparison.cdc
+
+Type<Int>() == Type<Int>()
+
+Type<Int>() != Type<String>()
 ```
 
-### `assert`
+The function `fun isInstance(_ type: Type): Bool` can be used to check if a value has a certain type.
 
 ```cadence
-fun assert(_ condition: Bool, message: String)
+let collectible <- create Collectible()
+let collectibleType = Type<@Collectible>()
+let result = collectible.isInstance(collectibleType)
+```
+
+For example, this allows implementing a marketplace sale resource:
+
+```cadence,file=runtime-type-simple-sale.cdc
+pub resource SimpleSale {
```

Comment:
It would probably be good to have some code comments in this file to make sure it is clear

Reply:
Good point, commented it in c69ff0d